### ISSUE TITLE
Convert double date values bug

### ISSUE THREAD

**Issue by akordowski** [mganss/ExcelMapper#51]
When converting a cell **double date** value, the following code

https://github.com/mganss/ExcelMapper/blob/baa149f5dbf0221f4145a2f629b7e26c23fc0c94/ExcelMapper/ExcelMapper.cs#L567-L571

should return `DateTime.FromOADate(cell.NumericCellValue)` instead of `cell.DateCellValue`. The return of `cell.DateCellValue` causes an exception in the NPOI library. Here the Stack Trace:

```
Exception: Object reference not set to an instance of an object.

at NPOI.XSSF.UserModel.XSSFWorkbook.IsDate1904()
at NPOI.XSSF.UserModel.XSSFCell.get_DateCellValue()
at Ganss.Excel.ExcelMapper.GetCellValue(ICell cell, ColumnInfo targetColumn)
at Ganss.Excel.ExcelMapper.<Fetch>d__54.MoveNext()
```

**Comment by mganss** [comment]
That's a bug in NPOI that's been fixed in https://github.com/tonyqus/npoi/commit/4890e9a518d90d85dd33eff6bc1ab74724d754b0#diff-510c4aedf3c5552c745b56c82f75f5da but their master is way ahead of the last NuGet release 🙄 

Using `FromOADate()` isn't 100% compatible with Excel:
https://docs.microsoft.com/en-us/office/troubleshoot/excel/1900-and-1904-date-system
https://www.joelonsoftware.com/2006/06/16/my-first-billg-review/

I think the best will be to work around this by catching the exception and try to recover by using `FromOADate()`.

**Comment by akordowski** [comment]
Currently there is no way how I can use a workaround or catch the exeption. I could use the workaround with the `SetPropertyUsing()`method, when the default `cell.NumericCellValue` were passed into the method. But currently the `cell.DateCellValue` is passed, which caused the exception. And the `GetCellValue()` method is called directly by the Fetch method.

Do you have an idea how I can currently fix this issue?

**Comment by mganss** [comment]
I'm sorry, I meant for the workaround to be applied inside ExcelMapper. I'll add it later today. Unfortunately, this means for a file where this issue occurs, an exception will be thrown for every cell read, slowing down performance. Do you know another way to handle this?

**Comment by akordowski** [comment]
I think in the current situation, until NPOI releases a new NuGet package, the best would be to return the original `cell.NumericCellValue` instead of a try/catch. So I could use the `SetPropertyUsing()` to convert the **double date** value to **date**. What do you think?

**Comment by mganss** [comment]
This would result in an error for the regular case where `SetPropertyUsing` is unset.

**Comment by akordowski** [comment]
Then, as I proposed, using the `DateTime.FromOADate()` until a new version of NPOI is released. In this state the method always throw an exception.

**Comment by mganss** [comment]
It throws only for those Excel files that trigger the NPOI bug ("some computer generated xlsx files" according to the commit message here https://github.com/tonyqus/npoi/commit/4890e9a518d90d85dd33eff6bc1ab74724d754b0).

I've added a temporary workaround that uses a static flag. So if `DateCellValue` throws a NullReferenceException once, we'll switch to using `DateUtil.GetJavaDate()` permanently.

**Comment by akordowski** [comment]
Thank you very much. That helps me a lot!

**Comment by akordowski** [comment]
And yes, it’s indeed a generated file that I’m using.

**Comment by mganss** [comment]
The bug fix in NPOI was released in 2.5.1.

**Comment by sachinjegaonkar** [comment]
Hi @mganss - I am still getting this error using ExcelMapper when reading computer generated xlsx files having some of the cell values as datetime. The ExcelMapper version that I am using is 5.0.166. When I publish my binaries, it shows NPOI version as 2.5.1.0. ExcelMapper v5.0.166 shows dependency on NPOI (>= 2.5.1). So, based on the above discussions, I was hoping the NullReferenceException should have got fixed in 5.0.166, but I am still getting this error. Please help me to get rid of this issue.

**Comment by mganss** [comment]
@sachinjegaonkar Please post a demo file.

**Comment by sachinjegaonkar** [comment]
> @sachinjegaonkar Please post a demo file.

Please find the attached demo file [demo.xlsx](https://github.com/mganss/ExcelMapper/files/4967839/demo.xlsx). Most of the times it reads in first attempt. But, for the subsequent attempts I get NullReferenceException for the same file.

`var objList = new ExcelMapper(excelFile) { HeaderRow = true }.Fetch<ObjModel>().ToList();`

This is the line of code I am using to map this demo excel into objList where all the excel headers are defined as string properties in ObjModel class along with the column attribute having names mapped as per the column header names.

**Comment by skokirala** [comment]
Hi,

I am also facing the same issue. I am using 2.5.1 version of NPOI. I am reading a Excel file from Blob storage as "Stream", which is able to read the date field for the first time, but it is throwing null reference exception for the subsequent requests. 

`var xssfwb = new XSSFWorkbook(excelStream);
`
`if (cell.CellType == CellType.Numeric)
                                    {
                                        if (DateUtil.IsCellDateFormatted(cell))
                                        {
                                            try
                                            {
                                                DateTime date = cell.DateCellValue;
                                                rowData[j] = date.ToShortDateString();
                                            }
                                            catch (NullReferenceException)
                                            {
                                                rowData[j] = cell.NumericCellValue.ToString();
                                            }
                                        }
                                        else
                                        {
                                            rowData[j] = cell.NumericCellValue.ToString();
                                        }

                                    }`

**Comment by sachinjegaonkar** [comment]
There seems to be an issue in ExcelMapper.cs in GetCellValue function at line #854 as below. For the second attempt (and onward) of reading excel file, cell.DateCellValue throws the NullReferenceException.
![image](https://user-images.githubusercontent.com/9928300/88389668-8a94f480-cdd4-11ea-85e3-96148b52d460.png)

Going further, it calls NPOI's DateUtil.GetJavaDate(cell.NumericCellValue) in catch block. Where that function also throws the NullReferenceException in [NPOI.SS.UserModel.DateUtil.GetJavaCalendar function at line #400](https://github.com/tonyqus/npoi/blob/2.5.1/main/SS/UserModel/DateUtil.cs)

![image](https://user-images.githubusercontent.com/9928300/88390317-bc5a8b00-cdd5-11ea-868a-13a20943efe6.png)


I have been able to proceed with the following workaround in ExcelMapper.cs file in GetCellValue function.

![image](https://user-images.githubusercontent.com/9928300/88390956-e19bc900-cdd6-11ea-84d5-271579bc3d62.png)

**Comment by mganss** [comment]
I can't repro using the demo file provided by @sachinjegaonkar. Please also post the class definition. What's your target framework?

**Comment by sachinjegaonkar** [comment]
> I can't repro using the demo file provided by @sachinjegaonkar. Please also post the class definition. What's your target framework?

My target framework is .NET Core 2.1 and here is the [class definition](https://github.com/mganss/ExcelMapper/files/4975652/ObjModel.txt) for the earlier shared demo file. Do you get any clue looking at the debug warning shown in [first screenshot above](https://user-images.githubusercontent.com/9928300/88389668-8a94f480-cdd4-11ea-85e3-96148b52d460.png) _**"The value of this expression may be incorrect. It could not be evaluated because: 'The function evaluation requires all threads to run."**_.

More information if that gives some clue - I am reading an excel file using ExcelMapper in one of the static methods of the class which is being called in different threads but in serialized way.

**Comment by mganss** [comment]
@sachinjegaonkar I'm sorry, I still can't repro. I've made a fresh .NET Core 2.1 console project and used your class definition as-is plus the `AddMapping` line for every property. Can you provide a self contained demo project that shows the issue?

**Comment by mganss** [comment]
@sachinjegaonkar @skokirala  Please check if this issue has been resolved in 5.1.180 (see discussion at #74).

**Comment by mganss** [comment]
Should be fixed for good in 5.1.294 which uses NPOI 2.5.3.
